From b0501258beaa6d95a1fae172e5f4ef4c4bdd962d Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Mon, 2 Dec 2024 15:53:36 +0100
Subject: [PATCH 1390/1695] clk: stm32mp21: check RIF access for HSI/HSE/MSI
 Kernel clocks

Check RIF access on clock enable/disable for HSI/HSE/MSI Kernel clocks.

Alignment with sTM32MP25 patch reviewed on
https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/412291

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git bcfde13478ca25a2d869d3b6b903721afc285f79 clk: stm32mp21: check RIF access for HSI/HSE/MSI Kernel clocks]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I4bc9680ace183831dfe9d45d29564f41edcab198
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/422151
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/clk/clk-stm32mp21.c | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/core/drivers/clk/clk-stm32mp21.c b/core/drivers/clk/clk-stm32mp21.c
index be77b91ac..1c007f953 100644
--- a/core/drivers/clk/clk-stm32mp21.c
+++ b/core/drivers/clk/clk-stm32mp21.c
@@ -2319,6 +2319,26 @@ static const struct clk_ops clk_stm32_oscillator_msi_ops = {
 static const struct clk_ops clk_stm32_no_ops = {
 };
 
+static TEE_Result clk_stm32_osc_ker_enable(struct clk *clk)
+{
+	if (stm32_rcc_has_access_by_id(RCC_RIF_OSCILLATORS))
+		return clk_stm32_gate_enable(clk);
+
+	return TEE_SUCCESS;
+}
+
+static void clk_stm32_osc_ker_disable(struct clk *clk)
+{
+	if (stm32_rcc_has_access_by_id(RCC_RIF_OSCILLATORS))
+		clk_stm32_gate_disable(clk);
+}
+
+static const struct clk_ops clk_stm32_osc_ker_ops = {
+	.enable		= clk_stm32_osc_ker_enable,
+	.disable	= clk_stm32_osc_ker_disable,
+	.is_enabled	= clk_stm32_gate_is_enabled,
+};
+
 static TEE_Result clk_stm32_hse_div_set_rate(struct clk *clk,
 					     unsigned long rate,
 					     unsigned long parent_rate)
@@ -3198,7 +3218,7 @@ static const struct clk_ops ck_timer_ops = {
 
 #define STM32_OSC_KERON(_name, _parent, _gate_id)\
 	struct clk _name = {\
-		.ops = &clk_stm32_gate_ops,\
+		.ops = &clk_stm32_osc_ker_ops,\
 		.priv = &(struct clk_stm32_gate_cfg){\
 			.gate_id = (_gate_id),\
 		},\
-- 
2.47.2

